sched/hrtimer: introduce scheduler support with hrtimer

wangchdo opened the pull request:
## Summary

This PR introduces high-resolution timer (hrtimer) support as a fully independent and optional module for support of the scheduler, without affecting existing scheduler behavior.

**Hrtimer is strictly isolated from the current scheduling logic:**

1. All hrtimer code is enabled only when CONFIG_HRTIMER is set.
2. If CONFIG_HRTIMER is disabled, the scheduler continues to use the existing tick-based or tickless mechanisms with no changes.
3. When enabled, hrtimer reuses the existing scheduler logic to drive OS ticks via a dedicated hrtimer instance.

**The module does not modify any scheduler data structures or timing paths.
Hrtimer acts solely as an alternative time source. Core scheduler functions (nxsched_process_tick(), nxsched_tick_expiration(), etc.) remain unchanged and are reused as-is.**

**Additional safeguards:**

1. The scheduler hrtimer is initialized lazily and does not impact system startup.
2. Existing timer infrastructure is not replaced or bypassed unless explicitly configured.
3. Any failure or imperfection in hrtimer is confined to the module itself and cannot affect legacy scheduling paths.

**Integration benefit**

This design enables incremental development and review of hrtimer while ensuring that existing NuttX scheduling behavior remains stable even if the hrtimer feature is explicitly enabled.

**Development benefit**
With this design, developers interested in optimizing the scheduler and those focused on optimizing hrtimer can work independently on their respective improvements.


## Impact

Add hrtimer support to nuttx scheduelr, without altering the existing scheduler behavior.

## Testing

**Test 1 passed (integrated in ostest):**

**- test implementation:**

```
/****************************************************************************
 * apps/testing/ostest/hrtimer.c
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.  The
 * ASF licenses this file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 ****************************************************************************/

/****************************************************************************
 * Included Files
 ****************************************************************************/

#include <nuttx/config.h>
#include <nuttx/hrtimer.h>

#include <stdio.h>
#include <sched.h>

#include "ostest.h"

/****************************************************************************
 * Pre-processor Definitions
 ****************************************************************************/

#define NSEC_PER_500MS (500 * NSEC_PER_MSEC)

/* Set a 1ms margin to allow hrtimertest to pass in QEMU.
 *
 * QEMU is a virtual platform, and its timer resolution and scheduling
 * latency may be less precise than on real hardware. Using a larger
 * margin ensures that tests do not fail due to timing inaccuracies.
 *
 * On real hardware (verified on the a2g-tc397-5v-tft board), this
 * margin can be reduced to less than 5 ns because timers are precise
 * and deterministic.
 */

#define NSEC_MARGIN    (NSEC_PER_MSEC)

/* Simple assertion macro for HRTimer test cases */
#define HRTIMER_TEST(expr, value)                                   \
  do                                                                \
    {                                                               \
      ret = (expr);                                                 \
      if (ret != (value))                                           \
        {                                                           \
          printf("ERROR: HRTimer test failed, line=%d ret=%d\n",   \
                 __LINE__, ret);                                    \
          ASSERT(false);                                            \
        }                                                           \
    }                                                               \
  while (0)

/****************************************************************************
 * Private Types
 ****************************************************************************/

/* Structure for HRTimer test tracking */

struct hrtimer_test_s
{
  hrtimer_t timer;    /* HRTimer instance */
  uint64_t  previous; /* Previous timestamp in nanoseconds */
  uint32_t  count;    /* Number of timer expirations */
  uint32_t  period;   /* Expected period between expirations */
  bool      active;   /* True while the test is still running */
};

/****************************************************************************
 * Private Functions
 ****************************************************************************/

/****************************************************************************
 * Name: hrtimer_test_init
 *
 * Description:
 *   Initialize a hrtimer_test_s structure for a new test.
 *
 * Input Parameters:
 *   test_hrtimer - Pointer to the test structure to initialize
 *   period       - Expected timer period in nanoseconds
 *
 * Returned Value:
 *   None
 *
 ****************************************************************************/

static void hrtimer_test_init(FAR struct hrtimer_test_s *test_hrtimer,
                              uint32_t period)
{
  test_hrtimer->previous = 0;
  test_hrtimer->count    = 0;
  test_hrtimer->active   = true;
  test_hrtimer->period   = period;
}

/****************************************************************************
 * Name: test_hrtimer_callback
 *
 * Description:
 *   HRTimer callback function for test.
 *
 *   - Verifies the timer interval is exactly 500ms (nanosecond precision)
 *   - Stops the test after 15 expirations
 *   - Re-arms the timer in absolute mode
 *
 * Input Parameters:
 *   hrtimer - Pointer to the expired HRTimer instance
 *
 * Returned Value:
 *   Timer period in nanoseconds (NSEC_PER_500MS)
 *
 ****************************************************************************/

static uint32_t test_hrtimer_callback(FAR hrtimer_t *hrtimer)
{
  struct timespec ts;
  uint32_t diff;
  uint64_t now;
  int ret;

  FAR struct hrtimer_test_s *test =
    (FAR struct hrtimer_test_s *)hrtimer;

  /* Increment expiration count */

  test->count++;

  /* Get current system time */

  clock_systime_timespec(&ts);
  now = clock_time2nsec(&ts);

  /* Skip comparison for first two invocations */

  if (test->count > 2)
    {
      /* Verify the timer interval is exactly
       * 500ms with nsec resolution
       */

      diff = (uint32_t)(now - test->previous);

      HRTIMER_TEST(NSEC_PER_500MS < diff + NSEC_MARGIN, true);
      HRTIMER_TEST(NSEC_PER_500MS > diff - NSEC_MARGIN, true);
    }

  test->previous = now;

  /* Stop the test after 15 expirations */

  if (test->count  >= 15)
    {
      ret = hrtimer_cancel(hrtimer);
      HRTIMER_TEST(ret, 0);

      test->active = false;
    }

  return test->period;
}

/****************************************************************************
 * Public Functions
 ****************************************************************************/

/****************************************************************************
 * Name: hrtimer_test
 *
 * Description:
 *   Entry point for high-resolution timer functional test.
 *
 *   - Initializes a HRTimer
 *   - Starts it with a 500ms relative timeout
 *   - Verifies subsequent expirations occur at 500ms intervals
 *
 * Input Parameters:
 *   None
 *
 * Returned Value:
 *   None
 *
 ****************************************************************************/

void hrtimer_test(void)
{
  int ret;
  struct hrtimer_test_s test_hrtimer_500ms;

  /* Initialize test structure */

  hrtimer_test_init(&test_hrtimer_500ms, NSEC_PER_500MS);

  /* Initialize the high-resolution timer */

  hrtimer_init(&test_hrtimer_500ms.timer,
               test_hrtimer_callback,
               NULL);

  /* Start the timer with 500ms relative timeout */

  ret = hrtimer_start(&test_hrtimer_500ms.timer,
                      test_hrtimer_500ms.period,
                      HRTIMER_MODE_REL);

  HRTIMER_TEST(ret, OK);

  /* Wait until the test completes */

  while (test_hrtimer_500ms.active)
    {
      usleep(500 * USEC_PER_MSEC);
    }
}
```

**test log on rv-virt:smp64:**


```
NuttShell (NSH)
nsh> 
nsh> uname -a
NuttX 0.0.0 6847a0cc95-dirty Dec 20 2025 12:26:39 risc-v rv-virt
nsh> ostest

(...)

user_main: hrtimer test

End of test memory usage:
VARIABLE  BEFORE   AFTER
======== ======== ========
arena     1fbdec0  1fbdec0
ordblks         9        9
mxordblk  1f73880  1f73880
uordblks    109b0     ff18
fordblks  1fad510  1fadfa8

Final memory usage:
VARIABLE  BEFORE   AFTER
======== ======== ========
arena     1fbdec0  1fbdec0
ordblks         1        9
mxordblk  1fb2cf8  1f73880
uordblks     b1c8     ff18
fordblks  1fb2cf8  1fadfa8
user_main: Exiting
ostest_main: Exiting with status 0
```

**test 2 passed (provided by @Fix-Point )**

**test implementation**

```
/****************************************************************************
 * apps/examples/hello/hello_main.c
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.  The
 * ASF licenses this file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 ****************************************************************************/

/****************************************************************************
 * Included Files
 ****************************************************************************/

#include <nuttx/config.h>
#include <stdio.h>
#include <unistd.h>

#include <nuttx/hrtimer.h>

#define HRTIMER_TEST_THREAD_NR (1)
#define HRTIMER_TEST_NR        (1000000)

/****************************************************************************
 * Public Functions
 ****************************************************************************/

static int volatile tcount = 0;
static volatile uint32_t next = 0;

static uint32_t test_callback(FAR struct hrtimer_s *hrtimer)
{

  tcount++;
  up_ndelay(hrtimer->expired % (10 * NSEC_PER_USEC));

  return 0;
}

static uint32_t test_callback_background(FAR struct hrtimer_s *hrtimer)
{
  up_ndelay(hrtimer->expired % NSEC_PER_USEC);
  return 0;
}


static void test1(int tid)
{
  hrtimer_t   timer;
  int         count = 0;
  irqstate_t flags;
  spinlock_t lock;

  if (tid == 0)
    {
      hrtimer_init(&timer, test_callback, NULL);
    }
  else
    {
      hrtimer_init(&timer, test_callback_background, NULL);
    }

  while (count++ < HRTIMER_TEST_NR)
    {
      int ret;
      if (tid == 0)
        {
          uint64_t delay = rand() % (10 * NSEC_PER_MSEC);

          /* Simulate the periodical hrtimer.. */

          flags = spin_lock_irqsave(&lock);

          /* Use as periodical timer */

          ret = hrtimer_cancel(&timer);
          ret = hrtimer_start(&timer, 1000, HRTIMER_MODE_REL);

          spin_unlock_irqrestore(&lock, flags);

          up_ndelay(NSEC_PER_MSEC);

          flags = spin_lock_irqsave(&lock);

          ret = hrtimer_cancel_sync(&timer);
          ret = hrtimer_start(&timer, 1000, HRTIMER_MODE_REL);
          spin_unlock_irqrestore(&lock, flags);

          up_ndelay(NSEC_PER_MSEC);

          hrtimer_cancel_sync(&timer); // stucked here????
          printf("???\n");
        }
      else
        {
          /* Simulate the background hrtimer.. */

          uint64_t delay = rand() % (10 * NSEC_PER_MSEC);

          ret = hrtimer_cancel(&timer);
          ret = hrtimer_start(&timer, delay, HRTIMER_MODE_REL);
        }

      UNUSED(ret);
    }
}

static void* test_thread(void *arg)
{
  while (1)
    {
      test1((int)arg);
    }
  return NULL;
}
/****************************************************************************
 * hello_main
 ****************************************************************************/

int main(int argc, FAR char *argv[])
{
  unsigned int   thread_id;
  pthread_attr_t attr;
  pthread_t      pthreads[HRTIMER_TEST_THREAD_NR];

  printf("hrtimer_test start...\n");

  ASSERT(pthread_attr_init(&attr) == 0);

  /* Create wdog test thread */

  for (thread_id = 0; thread_id < HRTIMER_TEST_THREAD_NR; thread_id++)
    {
      ASSERT(pthread_create(&pthreads[thread_id], &attr,
                            test_thread, (void *)thread_id) == 0);
    }

  for (thread_id = 0; thread_id < HRTIMER_TEST_THREAD_NR; thread_id++)
    {
      pthread_join(pthreads[thread_id], NULL);
    }

  ASSERT(pthread_attr_destroy(&attr) == 0);

  printf("hrtimer_test end...\n");
  return 0;
}
```

**test passed log on rv-virt:smp64**

```
nsh> uname -a
NuttX 0.0.0 6847a0cc95-dirty Dec 20 2025 12:26:39 risc-v rv-virt
nsh> 
nsh> hello
???
???
???
(...)
```

**test 3 passed (provided by @Fix-Point )**

**test implementation**

```
/****************************************************************************
 * apps/examples/hello/hello_main.c
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.  The
 * ASF licenses this file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 ****************************************************************************/

/****************************************************************************
 * Included Files
 ****************************************************************************/

#include <nuttx/config.h>
#include <stdio.h>

#include <nuttx/hrtimer.h>

#define HRTIMER_TEST_THREAD_NR (CONFIG_SMP_NCPUS * 8)

/****************************************************************************
 * Public Functions
 ****************************************************************************/

static uint32_t test_callback(FAR struct hrtimer_s *hrtimer)
{
  printf("test callback ...\n");
  return 0;
}

static void* test_thread(void *arg)
{
  irqstate_t  flags;
  hrtimer_t   timer;
  spinlock_t  lock = SP_UNLOCKED;
  hrtimer_init(&timer, test_callback, NULL);
  while (1)
    {
      uint64_t delay = rand() % NSEC_PER_MSEC;
      int ret;

      /* Simulate the usage of driver->wait_dog. */

      flags = spin_lock_irqsave(&lock);

      /* The driver lock acquired */

      /* First try, failed. Because hrtimer_start can not ensure the timer being started. */

      ret = hrtimer_cancel(&timer);
      // ret = hrtimer_start(&timer, 10 * NSEC_PER_USEC, HRTIMER_MODE_REL); /* May fail */

      /* This try-loop start should be OK. But it failed again.
       * Besides, we can not sleep or spin in the critical sections.
       */

      while (hrtimer_start(&timer, 10 * NSEC_PER_USEC, HRTIMER_MODE_REL) != OK);
      ret = OK;

      /* Second try, Success, but we can not sleep or spin in the critical section. */

      // ret = hrtimer_cancel_sync(&timer); /* Sleep in critical sections */
      // ret = hrtimer_start(&timer, delay, HRTIMER_MODE_REL);


      spin_unlock_irqrestore(&lock, flags);

      if (ret != OK)
        {
          printf("hrtimer_start failed\n");
        }
      up_ndelay(delay);
    }
  return NULL;
}

/****************************************************************************
 * hello_main
 ****************************************************************************/

int main(int argc, FAR char *argv[])
{
  unsigned int   thread_id;
  pthread_attr_t attr;
  pthread_t      pthreads[HRTIMER_TEST_THREAD_NR];

  printf("hrtimer_test start...\n");

  ASSERT(pthread_attr_init(&attr) == 0);

  /* Create wdog test thread */

  for (thread_id = 0; thread_id < HRTIMER_TEST_THREAD_NR; thread_id++)
    {
      ASSERT(pthread_create(&pthreads[thread_id], &attr,
                            test_thread, NULL) == 0);
    }

  for (thread_id = 0; thread_id < HRTIMER_TEST_THREAD_NR; thread_id++)
    {
      pthread_join(pthreads[thread_id], NULL);
    }

  ASSERT(pthread_attr_destroy(&attr) == 0);

  printf("hrtimer_test end...\n");
  return 0;
}

```

**test passed log on rv-virt:smp64**

```
NuttShell (NSH)
nsh> 
nsh> 
nsh> uname -a
NuttX 0.0.0 6847a0cc95-dirty Dec 20 2025 12:33:59 risc-v rv-virt
nsh> hello
test callback ...
test callback ...
test callback ...
test callback ...
test callback ...
test callback ...
test callback ...
test callback ...
test callback ...
test callback ...
test callback ...

(....)
```

wangchdo commented:
@cederom @acassis @anchao 

Conflicts resolved, please double check

wangchdo commented:
Just fixed CI issue, please help to double check @acassis @anchao @cederom

anchao commented on sched/hrtimer/hrtimer_process.c:
```diff
@@ -100,27 +99,81 @@ void hrtimer_process(uint64_t now)
+                      hrtimer->state = HRTIMER_STATE_INACTIVE;
+#ifdef CONFIG_SMP
+                    }
+#endif
```
```suggestion
#ifdef CONFIG_SMP
                  if (hrtimer->cpus == 0)
#endif
                    {
                      hrtimer->state = HRTIMER_STATE_INACTIVE;
                    }
```

anchao commented on sched/hrtimer/hrtimer_process.c:
```diff
@@ -100,27 +99,81 @@ void hrtimer_process(uint64_t now)
+            {
+              /* Timer was canceled during callback execution */
+
+#ifdef CONFIG_SMP
```
ditto

anchao commented on sched/sched/sched_processtimer.c:
```diff
@@ -44,11 +44,139 @@
+
+static void nxsched_process_hrtimer(void)
+{
+  uint64_t now = hrtimer_gettime();
```
move hrtimer_gettime() before hrtimer_process()

wangchdo replied:
Done

Fix-Point commented:
@anchao @wangchdo  I have serious doubts about your professionalism.  Why are you allowing such a functional incorrect code to be merged into the master branch. Isn’t this simply a case of factional bias clouding judgment? 
Please check this, thanks.
https://github.com/apache/nuttx/pull/17675#issuecomment-3698712790

xiaoxiang781216 commented on include/nuttx/hrtimer.h:
```diff
@@ -102,9 +102,7 @@ struct hrtimer_s
   hrtimer_cb func;             /* Expiration callback function */
   FAR void *arg;               /* Argument passed to callback */
   uint64_t expired;            /* Absolute expiration time (ns) */
-#ifdef CONFIG_SMP
```
remove this patch from pr

xiaoxiang781216 commented on drivers/timers/arch_alarm.c:
```diff
@@ -118,8 +118,10 @@ static void ndelay_accurate(unsigned long nanoseconds)
-#ifdef CONFIG_SCHED_TICKLESS
-  nxsched_timer_expiration();
+#if defined(CONFIG_HRTIMER) || \
+    defined(CONFIG_SCHED_TICKLESS)
```
merge to one line

wangchdo commented on sched/hrtimer/hrtimer_cancel.c:
```diff
@@ -27,6 +27,7 @@
 #include <nuttx/config.h>
 #include <nuttx/arch.h>
 #include <nuttx/clock.h>
+#include "sched/sched.h"
```
Done

Fix-Point commented:
> > @anchao @wangchdo I have serious doubts about your professionalism.
> > > @wangchdo's HRTimer implementation has **fundamental functional correctness issues. It violates the ownership invariant of hrtimer**: that only one reader/writer can use the hrtimer object at a time. This violation makes it prone to issues in SMP scenarios, as I have consistently pointed out over and over again. Simple modifications like reference counting can not fix this implementation at all.
> > 
> > 
> > Why are you allowing such a functional incorrect code to be merged into the master branch. Isn’t this simply a case of factional bias clouding judgment? Please check this, thanks. [#17675 (comment)](https://github.com/apache/nuttx/pull/17675#issuecomment-3698712790)
> 
> @Fix-Point This PR addresses the SMP corner cases you mentioned. Could you please take a look? I propose this solution for the following reasons:
> 
> 1. It allows separate optimization paths for SMP and non-SMP, leaving room for extreme performance and strong determinism in both cases. (per @xiaoxiang781216 's comments I will submit a separate PR to put the SMP specific logics in CONFIG_SMP )
> 2. It provides clean, easy-to-use and  minimal user-facing APIs — only three core APIs, with no duplicated parameters:
>    `hrtimer_init()`
>    `hrtimer_start()`
>    `hrtimer_cancel() / hrtimer_cancel_sync()`
> 3. It provides basement for queue abstraction if needed.
> 
> @xiaoxiang781216 I resolved you comments, could you please take a look again @acassis @anchao please double check

To behonest, you've only modified the code to prevent the test program from getting stuck there. You haven't solved the actual problem: **canceled periodic timer cannot overwrite the new timer**.

I've emphasized this many times. Why make futile attempts again and again? Do you know how much of our time you've wasted?

>More similar concurrency issues could be cited here. As I have emphasized again and again, **the fundamental problem is the violation of the ownership invariant of hrtimer: only one owner can modify the hrtimer object at a time**.
**Designing functionally correct concurrent algorithms is not easy at all. Relying solely on engineering experience is insufficient; theoretical methods are necessary to avoid errors, such as adapting resource invariants and using structured diagrams to clarify every possible concurrent state transition**. @wangchdo's design failed to consider how to handle concurrency correctly, making it nearly impossible to improve upon his code base.

wangchdo commented on sched/hrtimer/hrtimer_process.c:
```diff
@@ -99,28 +98,70 @@ void hrtimer_process(uint64_t now)
+
+      /* Execute the timer callback */
 
       period = hrtimer->func(hrtimer);
```
I think the solution is straightforward: if you want to start an hrtimer with new parameters, make sure to do so only after calling hrtimer_cancel_sync().

This approach makes hrtimer processing deterministic and, at the same time, improves the efficiency of hrtimer handling.

wangchdo commented:
> @wangchdo please rebase with current mainline to fix the CI error

@acassis 
Rebased, please double check

wangchdo commented:
@cederom  CI passed, could you please double check?

@xiaoxiang781216 @Fix-Point
I’ve addressed your comments—could you please take another look?

I understand that your main concern with my previous implementation was about potential concurrency issues, or a possible violation of ownership invariants under SMP. This PR resolves those concerns using an approach that I believe is **reliable, simple, and efficient**. Below, I’ll explain the fix from a code-level perspective:

1. In `hrtimer_process()`, I now copy the `hrtimer callback`, `expiration time`, and `argument` into local variables while holding the spinlock:

![img_v3_02ti_fc18d9f4-f5dd-4992-bd20-782b39ea245g](https://github.com/user-attachments/assets/dbe73bee-8da7-4afd-bf63-9f48d20e3ed1)

2. Still in hrtimer_process(), the callback is invoked after the spinlock is released, using those local variables. This makes it safe for other cores to update the same hrtimer’s callback or expiration concurrently via `hrtimer_set()`, `hrtimer_cancel()`, or `hrtimer_start()`:

![img_v3_02ti_b4096580-0945-44f3-959c-6e254a350d6g](https://github.com/user-attachments/assets/67345bd8-9b21-4ac8-89e6-02863dfa36d6)

3. In `hrtimer_cancel()`, `hrtimer->expired` is set to `UINT64_MAX` to explicitly mark the hrtimer as cancelled:

![img_v3_02ti_aa7370d2-2eb5-4cd4-8821-6b75795abbag](https://github.com/user-attachments/assets/662a03ad-6176-4380-8278-9772b3f46ba2)

4 In `hrtimer_start()`, `hrtimer->expired` is updated directly with the new expiration value:

![img_v3_02ti_b6193a60-23f6-4e4b-a318-e067be9a99eg](https://github.com/user-attachments/assets/7d1f613e-a779-46ab-adcb-6e06ab579fc9)

5. Also in `hrtimer_process()`, if period > 0, the hrtimer is only allowed to restart when `hrtimer->expired` has not been modified concurrently by other cores. Both `hrtimer_cancel() `(which sets expired to UINT64_MAX) and `hrtimer_start() `(which sets it to a new value) will cause this check to fail. This guarantees that the **hrtimer will not be restarted unexpectedly**:

![img_v3_02ti_b98e97a2-d688-4b1d-a072-86a4719be23g](https://github.com/user-attachments/assets/68e427a0-78ac-412e-b044-ec850056303b)

6. Finally, in` hrtimer_process()`, `hrtimer->cpus `acts as a reference counter and is automatically incremented and decremented. This allows it to be safely used by `hrtimer_cancel_sync() `to ensure the hrtimer is fully quiesced before returning:

![img_v3_02ti_a2050e68-d10c-4298-868b-7977937ef5dg](https://github.com/user-attachments/assets/662cb092-5980-47be-9aa7-580b6f75b270)
![img_v3_02ti_8f12799e-ce5e-432f-8dec-1c50d96eeceg](https://github.com/user-attachments/assets/e128790e-74ef-4a4e-87f8-4ffbd000cbab)

**Overall, this design ensures correct behavior under SMP while keeping the logic straightforward and efficient, and it integrates cleanly with the existing hrtimer model in NuttX.**

wangchdo commented:
@Fix-Point @xiaoxiang781216 

As I pointed out in the comments on your PR (#17675), I don’t think the new hrtimer implementation is a good choice.

From my perspective, the way it addresses concurrency issues is not efficient, and it requires updating the entire hrtimer API surface, which I believe makes it less user-friendly compared to my approach.

My implementation focuses on resolving the concurrency concerns without forcing API-wide changes, keeping the usage model simple while still ensuring correctness under SMP.

wangchdo commented on sched/hrtimer/hrtimer_process.c:
```diff
@@ -94,32 +106,36 @@ void hrtimer_process(uint64_t now)
+
+      /* Re-arm periodic timer if not canceled or re-armed concurrently */
+
+      if (period > 0 && hrtimer->expired == expired)
```
> I still insist what I stated before:
> 
> > Designing correct concurrent algorithms requires systematic consideration. As I stated in #17570, after carefully reviewing your implementation, I could not really find a good solution that preserves version information without introducing significant performance or memory overhead.

I think your new implementation is not proper due to three reasons:

1. Your way of protecting timer from concurrency update is not efficient

2. You totally refactored the existing more user-friendly API design, this is not correct

3. Your queue abstraction design should be based on the existing hrtimer design.in fact, if you do not provide your way of queue abstraction, i already planned to add mine.And, In my opinion, maybe I am wrong,Nuttx is a RTOS, not a OS as linux, but I found your queue abstraction has a Linux style.

wangchdo commented:
> > Thank you @wangchdo! This change builds fine now. @Fix-Point provides some important feedback as he seems to work on this already. I like approach of @wangchdo to keep things compatible and aligned with existing API. Maybe additional checks / protections may be implemented to avoid situations described by @Fix-Point? :-)
> 
> > 
> 
> > @Fix-Point can you please provide exact test scenario steps to verify problems you mentioned on @wangchdo solution? This should confirm if the implementation requires some additional protections? :-)
> 
> 
> 
> I have provided @wangchdo  with 4-5 test cases and believe I have been very patient in explaining why this approach will not work. However, he still refuses to accept my reasoning.
> 
> 
> 
> Here is the detailed explanation:
> 
> 
> 
> First, to ensure concurrency correctness, we must adhere to resource invariants, such as the ownership invariant. That is, only one user can update a shared object at any given time.
> 
> 
> 
> In the hrtimer design, there may be situations where a new hrtimer is **already being reset while the callback function of the old hrtimer is still executing**. For such scenarios, we must ensure in the hrtimer design that "a cancelled timer cannot modify the hrtimer object again." In other words, the execution core of the old timer callback loses ownership of the hrtimer object. Otherwise, it could lead to the old timer overwriting the new timer, which is an error. This is why I refer to his implementation as "functionally incorrect" (i.e., it violates the specification).
> 
> 
> 
> **The key problem of its implementation is the violation of the ownership invariant, which is impossible to fix without introducing extra memory overhead**. Let me elaborate on why his latest design still violates the ownership invariant:  
> 
> After callback execution completes, a periodic timer requeues itself. In his latest implementation, the requeue condition is:
> 
> ```c
> 
> // executing callback...
> 
> if (hrtimer->expired != UINT64_MAX && hrtimer->expired != expired)
> 
> ```
> 
> It violates the ownership invariant when:
> 
> 1. If the callback function for the current `hrtimer->expired = t1` is executing and returns 0.
> 
> 2. While the callback is executing, the timer is cancelled and a new timer is started with `hrtimer->expired = t2`.
> 
> 3. When the callback finishes, it finds `hrtimer->expired != t1` and mistakenly requeue the one-shot timer as a periodic timer, causing a system crash.
> 
> 
> 
> 2. Even reverting to the original implementation still violates the invariant because **`expired` is not monotonically increasing and cannot be used as a version**. It is possible for the old and new timers to have the same `expired` time, allowing a cancelled timer to modify the new timer.
> 
> 
> 
> ```c
> 
> // executing callback...
> 
> if (period > 0 && hrtimer->expired == expired)
> 
> ```
> 
> 
> 
> It violates the ownership invariant when:
> 
> 1. If a periodic timer `hrtimer->expired = t1`, `hrtimer->func = period_func` is executing.
> 
> 2. While the callback is executing, the timer is cancleled and a new timer is started with `hrtimer->expired = t1`, `hrtimer->func = oneshot_func`.
> 
> 3. When the cancelled `period_func` finishes, while `oneshot_func` is still executing, the cancelled `period_func` overwrites the newly set oneshot timer, causing an error.
> 
> 
> 
> Even if we add a check for `func` as suggested by @wangchdo , problems remain:
> 
> ```c
> 
> // executing callback...
> 
> if (period > 0 && hrtimer->expired == expired && hrtimer->func == func)
> 
> ```
> 
> 
> 
> It violates the ownership invariant when:
> 
> 1. If a periodic timer `hrtimer->expired = t1`, `hrtimer->func = period_func` is executing, and `period_func` returns the next delay `p1`.
> 
> 2. At time `t1`, while the hrtimer callback is executing, the periodic timer is cancelled, and a new periodic timer is set to trigger immediately with `hrtimer->expired = t1`, `hrtimer->func = period_func`, which returns the next delay `p2`.
> 
> 3. When the cancelled previous callback finishes, while the new callback is still executing, all version checks pass, and the next timer is set to `hrtimer->expired = t1 + p1`, overwriting the new timer (the expected next expired time should be `hrtimer->expired = t1 + p2`).
> 
> 
> 
> The correct solution is to add a monotonically increasing version field, `version`, to the hrtimer. When an hrtimer is cancelled, `version` increments by 1, ensuring that **an cancelled old timer with old version cannot overwrite a newly set timer with newer version**. However, **this would inevitably increase the memory footprint of the `hrtimer`**. I considered this approach early in the design phase but ultimately opted for `hazard pointers` to avoid the memory overhead.
> 
> 
> 
> BTW, I feel that @wangchdo  has shown little respect for me:
> 
> 1.  I and @wangchdo each implemented our hrtimer independently, with almost no communication during the process. These implementations are completely different. In mid-December, @xiaoxiang781216  suggested that I share my design with @wangchdo . During our exchange, I pointed out his errors regarding SMP, but he refused to acknowledge them and accused me of stealing his ideas. He also asked me to submit my incomplete internal hrtimer implementation to the community.
> 
> 2.  Out of respect for his opinion, I submitted my design and demonstrated that it outperforms his in both functional correctness and performance. After I submitted my code implementation to the community, to my surprise, without any discussion, @anchao forcibly merged @wangchdo's functional incorrect implementation into the upstream, which made me feel very very uncomfortable. 
> 
> 4. Yet, @wangchdo  consistently refused to replace his implementation with mine. I cannot understand why he remains so stubborn.
> 
> 5. He rarely tests his own code; even pulling his PR and compiling it reveals compilation issues. He repeatedly pressured me to provide test cases for him but never offered any in return, which I find insulting.
> 
> 6. I have consistently tried to reason with him, presenting performance test results, interface comparisons, and memory usage analyses. However, he refuses to acknowledge them and has not provided any convincing evidence to support his stance.
> 
> 
> 
> I recommend that @wangchdo  refer to my design and adopt hazard pointers, which are an optimal solution for memory efficiency. I have no issue with him using my design consideration (in fact, hazard pointers is memory reclamation technique proposed by others). Similar design goals inevitably lead to convergent designs, and this is entirely natural.
> 
> 
> 
> Finally, Could you please help me review my implementation https://github.com/apache/nuttx/pull/17675? I welcome any feedback and believe that, at the very least, my implementation is superior to his in terms of functional correctness, design completeness, documentation, performance, scalability, and code reusability. =)
> 
> 
> 
> Regarding the APIs, I have made every effort to align it with the original design. However, some aspects are inherently tied to the design and difficult to change. For example, my implementation encodes the state in `hrtimer->func`, which means restarting the timer requires passing the `func`. Besides, according to test results, this does not impact performance.

@Fix-Point  This is my last response to you:

It is just about design choice 
when we decide how to handle concurrency issues. What is most important before you make a decision of whether the implementation is right, you only need to consider if the behavior is what we want. That is what is the requirement? 

In my opinion, there is no wrong implementation, there is only implementation that meets requirements and others no.

Anyway, I don't want to talk about anything else not related to technical stuff. And I don't want judging anyone personally happens in our community.

xiaoxiang781216 commented:
> 4. Even reverting to the original implementation still violates the invariant because **`expired` is not monotonically increasing and cannot be used as a version**. It is possible for the old and new timers to have the same `expired` time, allowing a cancelled timer to modify the new timer.
> 
> ```c
> // executing callback...
> if (period > 0 && hrtimer->expired == expired)
> ```
> 
> It violates the ownership invariant when:
> 
> 1. If a periodic timer `hrtimer->expired = t1`, `hrtimer->func = period_func` is executing.
> 2. While the callback is executing, the timer is cancleled and a new timer is started with `hrtimer->expired = t1`, `hrtimer->func = oneshot_func`.
> 3. When the cancelled `period_func` finishes, while `oneshot_func` is still executing, the cancelled `period_func` overwrites the newly set oneshot timer, causing an error.


but we just provide hrtimer_init to change `hrtimer->func`, which mean the `func` can't be changed after initialization.

> 
> Even if we add a check for `func` as suggested by @wangchdo , problems remain:
> 
> ```c
> // executing callback...
> if (period > 0 && hrtimer->expired == expired && hrtimer->func == func)
> ```
> 
> It violates the ownership invariant when:
> 
> 1. If a periodic timer `hrtimer->expired = t1`, `hrtimer->func = period_func` is executing, and `period_func` returns the next delay `p1`.
> 2. At time `t1`, while the hrtimer callback is executing, the periodic timer is cancelled, and a new periodic timer is set to trigger immediately with `hrtimer->expired = t1`, `hrtimer->func = period_func`, which returns the next delay `p2`.
> 3. When the cancelled previous callback finishes, while the new callback is still executing, all version checks pass, and the next timer is set to `hrtimer->expired = t1 + p1`, overwriting the new timer (the expected next expired time should be `hrtimer->expired = t1 + p2`).

in this case, the timeout will be set to `hrtimer->expired = t1 + p1`, but overwrite by `hrtimer->expired = t1 + p2`. I think it's right behaviour the last timeout win the competition.

xiaoxiang781216 commented on sched/sched/sched_processtimer.c:
```diff
@@ -44,11 +44,145 @@
+ * nxsched_process_hrtimer().
+ */
+
+static bool g_sched_hrtimer_inited = false;
```
remove, don't need with https://github.com/apache/nuttx/pull/17784

wangchdo replied:
Done

xiaoxiang781216 commented on include/nuttx/arch.h:
```diff
@@ -2031,7 +2031,8 @@ int up_alarm_tick_cancel(FAR clock_t *ticks);
  ****************************************************************************/
 
-#if defined(CONFIG_SCHED_TICKLESS) && defined(CONFIG_SCHED_TICKLESS_ALARM)
+#if (defined(CONFIG_HRTIMER) && defined(CONFIG_ALARM_ARCH)) || \
```
remove CONFIG_ALARM_ARCH

xiaoxiang781216 commented on include/nuttx/arch.h:
```diff
@@ -2101,8 +2102,8 @@ int up_timer_tick_cancel(FAR clock_t *ticks);
  ****************************************************************************/
-
-#if defined(CONFIG_SCHED_TICKLESS) && !defined(CONFIG_SCHED_TICKLESS_ALARM)
+#if (defined(CONFIG_HRTIMER) && defined(CONFIG_TIMER_ARCH)) || \
```
why change up_timer_start

xiaoxiang781216 commented on sched/sched/sched_processtimer.c:
```diff
@@ -44,11 +44,145 @@
+  /* Notify scheduler of the current time (converted to ticks) */
+
+  nxsched_tick_expiration(NSEC2TICK(now));
+
```
not return value

wangchdo commented on include/nuttx/arch.h:
```diff
@@ -2450,6 +2450,25 @@ void up_ndelay(unsigned long nanoseconds);
+ ****************************************************************************/
+
+#ifdef CONFIG_HRTIMER
+int nxsched_hrtimer_start(clock_t ticks);
```
Done

wangchdo commented:
> let's squash into one patch @wangchdo

OK, Done.

xiaoxiang781216 commented on include/nuttx/arch.h:
should we merge nxsched_timer_expiration and nxsched_process_timer into one first?

xiaoxiang781216 commented on sched/sched/sched_processtimer.c:
```diff
@@ -44,11 +44,108 @@
+ * CONFIG_HRTIMER is enabled but not used immediately.
+ */
+
+static hrtimer_t g_nxsched_hrtimer;
```
```
static hrtimer_t g_nxsched_hrtimer =
{
  nxsched_hrtimer_callback,
};
```

wangchdo replied:
Done

cederom commented on drivers/timers/arch_alarm.c:
```diff
@@ -118,8 +118,9 @@ static void ndelay_accurate(unsigned long nanoseconds)
-#ifdef CONFIG_SCHED_TICKLESS
-  nxsched_timer_expiration();
+#if defined(CONFIG_HRTIMER) || defined(CONFIG_SCHED_TICKLESS)
+  nxsched_process_timer();
```
> @xiaoxiang781216: should we update all driver and remove nxsched_timer_expiration

wouldn't that break backward compatibility with existing tickless timer? :-P

wangchdo commented on include/nuttx/arch.h:
```diff
@@ -2031,7 +2031,8 @@ int up_alarm_tick_cancel(FAR clock_t *ticks);
  ****************************************************************************/
 
-#if defined(CONFIG_SCHED_TICKLESS) && defined(CONFIG_SCHED_TICKLESS_ALARM)
+#if defined(CONFIG_HRTIMER)|| \
```
Done

wangchdo commented:
> @wangchdo could you please submit these hrtimer test to nuttx-apps/testing/sched/hrtimer ?

OK
